contracts/PrepaidCardMarket_v2.sol

```diff
@@ -0,0 +1,304 @@
+
+    return balance[issuerSafe][token] / price;
+  }
+
+  function setAsk(
+    address issuer,
+    bytes32 sku,
+    uint256 askPrice
+  ) external returns (bool) {
+    require(skus[sku].issuer != address(0), "Non-existent SKU");
+    require(skus[sku].issuer == issuer, "SKU not owned by issuer");
+    asks[sku] = askPrice;
+
+    emit AskSet(issuer, skus[sku].issuingToken, sku, askPrice);
+    return true;
+  }
+
+  function addSKU(
+    uint256 faceValue,
+    string memory customizationDID,
+    address token
+  ) external returns (bool) {
+    require(faceValue > 0, "Face value must be greater than 0");
+    require(token != address(0), "Token address must be set");
+
+    address _issuer = issuers[msg.sender];
+    require(_issuer != address(0), "Issuer not found.");
+
+    bytes32 sku = getSKU(_issuer, token, faceValue, customizationDID);
+
+    require(skus[sku].issuer == address(0), "SKU already exists");
+
+    skus[sku] = SKU({
+      issuerSafe: msg.sender,
+      issuer: _issuer,
+      issuingToken: token,
+      faceValue: faceValue,
+      customizationDID: customizationDID
+    });
+
+    emit SkuAdded(_issuer, token, faceValue, customizationDID, sku);
+
+    return true;
+  }
+
+  function getSKU(
+    address issuer,
+    address token,
+    uint256 faceValue,
+    string memory customizationDID
+  ) public pure returns (bytes32) {
+    return
+      keccak256(abi.encodePacked(issuer, token, faceValue, customizationDID));
+  }
+
+  function withdrawTokens(uint256 amount, address token) external {
+    address _issuer = issuers[msg.sender];
+    require(_issuer != address(0), "Issuer not found");
+
+    uint256 balanceForToken = balance[msg.sender][token];
+
+    require(amount <= balanceForToken, "Insufficient funds for withdrawal");
+
+    balance[msg.sender][token] -= amount;
+
+    IERC677(token).safeTransfer(msg.sender, amount);
+
+    emit TokensWithdrawn(msg.sender, _issuer, token, amount);
+  }
+
+  function onTokenTransfer(
+    address payable from, // safe address
+    uint256 amount,
+    bytes calldata data
+  ) external returns (bool) {
+    // Only CARD.CPXD, DAI.CPXD are accepted
+    require(
+      TokenManager(tokenManager).isValidToken(msg.sender),
+      "token is unaccepted"
+    );
+
+    address _issuer = abi.decode(data, (address));
+
+    require(_issuer != address(0), "issuer should be provided");
+
+    address[] memory _owners = GnosisSafe(from).getOwners();
+    bool _foundOwner = false;
+
+    // Caution: block gas limit could be too high for big arrays
+    require(_owners.length < 100, "too many safe owners");
+
+    for (uint256 i = 0; i < _owners.length; i++) {
+      if (_owners[i] == _issuer) {
+        _foundOwner = true;
+        break;
+      }
+    }
+
+    require(_foundOwner, "issuer is not one of the safe owners");
+
+    balance[from][msg.sender] = balance[from][msg.sender] + amount;
+    issuers[from] = _issuer;
```

Comment:
we could also add an `issuersSafe` that is the inverse of `issuers` here too in order to help with the `getQuantity` calculations